src/Symfony/Component/EventDispatcher/EventDispatcher.php

```diff
@@ -30,6 +30,14 @@ class EventDispatcher implements EventDispatcherInterface
 {
     private $listeners = array();
     private $sorted = array();
+    private $optimized;
+
+    public function __construct()
+    {
+        if (__CLASS__ === \get_class($this)) {
```

Comment:
because this optimization is internal: the `getListeners()` should continue returning the exact listeners that where added - doing otherwise would be a BC break

Reply:
getListeners() is currently completely detached from optimizeListeners()

Reply:
`dispatch()` calls `getListeners()`: in case the class is overridden, it should continue calling it.
This means we cannot call `optimizeListeners()` for child classes.

Reply:
Well that sucks. Can we deprecate calling `getListeners` from `dispatch` by passing special flag?

Reply:
Not in this PR at least :)